### ISSUE TITLE
[prefer-to-have-class] does not trigger for elements, only for DTS queries

### ISSUE THREAD

**Issue by danielrentz** [testing-library/eslint-plugin-jest-dom#490]
The rule `prefer-to-have-class` does not trigger when passing an element, but only for DTL queries:

```js
// this triggers as expected
expect(screen.getByTestId('id')).toHaveAttribute('class', 'test')

// this does not trigger
expect(document.body).toHaveAttribute('class', 'test')
```

According to the [documentation page](https://github.com/testing-library/eslint-plugin-jest-dom/blob/main/docs/rules/prefer-to-have-class.md#rule-details) this should work.

Seems the rule implementation only accepts DTL queries and nothing else.
https://github.com/testing-library/eslint-plugin-jest-dom/blob/main/src/rules/prefer-to-have-class.js#L238-L244

Same problem for other variants like accessing `classList`, `className` etc.

The rules `prefer-to-have-attribute`, `prefer-to-have-style`, `prefer-to-have-text-content` work as expected.

**Comment by G-Rath** [comment]
hmm interesting - it seems that #129 did this with no backing issue or reasoning.

However, it actually looks like most of our rules are implemented to only apply to DTL queries, except for 

- prefer-focus
- prefer-to-have-attribute
- prefer-to-have-style
- prefer-to-have-text-content

so arguably it's a bug that they don't?

I'm very interested in discussing this further and hearing other peoples opinions:

- afaik `jest-dom` is only connected to DTL through its org owner and maintainers, not by dependency or runtime
- not doing these checks would let us remove an optional peer dependency and decent amount of code
- most if not all of the properties I think are not super generic that they should cause tons of false positives

on the other hand this has been the behaviour for a while, so it could be argued changing it is breaking - but I think that really comes down to how many false positives there are 🤔 

I might try and play around with this on our codebases and the smoke testing infra to see if I can get a better sense of the impact of this change

**Comment by danielrentz** [comment]
Well there are pros and cons I think :)

In our pretty large codebase, we disabled the rule `prefer-to-have-class` a few months ago because it crashed with Eslint 10. Now after re-enabling the rule, it does not trigger anymore in a small "unit-test" like file used to check if the linter plugins work as expected (it provokes linter findings together with inline disable comments and the "reportUnusedDisableDirectives" flag). So it "worked" in the past and now it doesn't.

On the other hand, in the same codebase, we disabled the rule `prefer-to-have-value` because it triggered way too often on tests unrelated to DOM elements. I checked the [current implementation of this rule](https://github.com/testing-library/eslint-plugin-jest-dom/blob/main/src/rules/prefer-to-have-value.js#L33-L38), it is now strictly bound to `*ByRole` queries with "textbox" and "dropdown" roles.

So this is my opinion here:
- I would be fine with restricting the other rules to DTL queries as well for consistency. I like consistency.
- This restriction should clearly be mentioned in the documentation of all rules.
- In my eyes, `to-have-value` is a little too strict, but I understand the motivation to prevent FPs here. Same as above - documentation page needs update.

**Comment by G-Rath** [comment]
> Now after re-enabling the rule, it does not trigger anymore in a small "unit-test" like file used to check if the linter plugins work as expected (it provokes linter findings together with inline disable comments and the "reportUnusedDisableDirectives" flag). So it "worked" in the past and now it doesn't.

Not sure what that is about, but it shouldn't be anything new - the main changes since I've taken over as the primary maintainer have mostly been about ESLint v10 compatibility.

If you share more code + details here or on a new issue, happy to dig into that further in case it's a latent bug.

> On the other hand, in the same codebase, we disabled the rule prefer-to-have-value because it triggered way too often on tests unrelated to DOM elements.

I assume this is running the rule across your whole test suite, rather than being scoped to just DOM-based tests? I'd like people to be able to just "set and forget" but given the size of your codebase it would be interesting to know the ratio of false positives (I assume you've got maybe a backend/frontend monorepo type situation?)
